runbooks/how-to-load-cloudwatch-historical-data.md

```diff
@@ -0,0 +1,200 @@
+    --where partition_0 = '2022_(06-0903)'
+    --and month(from_iso8601_timestamp("timestamp")) <= 6   --  9m20s, 92.13GB
+    --where partition_0 = '2022_(06-0903)'
+    --and month(from_iso8601_timestamp("timestamp")) in (7,8) -- 16m05s, 92.13GB
+    where partition_0 = '2022_(06-0903)'                      --  4m08s, 92.46GB
+      and month(from_iso8601_timestamp("timestamp")) >= 9
+      and json_extract_scalar(message,'$.mdc["http_access_response_header_X-Request-ID"]') not in (
+            select "mdc_http_access_response_header_x-request-id"
+            from "bfd-insights-bfd-prod".bfd_insights_bfd_prod_api_requests
+      )
+   ) select
+      cw_timestamp,
+      cw_id,
+      timestamp,
+      level,
+      thread,
+      logger,
+      message,
+      context,
+      stuff['bene_id'] as "bene_id",
+      stuff['database_query_bene_by_coverage_batch'] as "database_query_bene_by_coverage_batch",
+      ... <repeat for all mdc columns in order since this is positional>
+      from dataset
+   ```
+
+6. Load the `target` table from the `staging` table.
+   1. The `target` table definition resides in terraform and should match the `staging` table columns and ordering with the one
+      difference being that the `staging` table does not include the `year` and `month` partition columns. Verify that
+      the table structure is the same if not done already. 
+   2. Similar to the loading of the `staging` table, this must be done in batches to avoid the Athena 30 minute timeout.
+      The batches here can be larger than for the `staging` table but ~500 million is the limit.
+      ```sql
+      insert into "bfd_insights_bfd_prod_api_requests" 
+      select
+         cw_timestamp,
+         cw_id,
+         timestamp,
+         level,
+         thread,
+         logger,
+         message,
+         context,
+         "mdc_bene_id",
+         "mdc_database_query_bene_by_coverage_batch",
+         "mdc_jpa_query_eobs_by_bene_id_snf_record_count",
+         ... <all mdc columns here>
+         date_format(from_iso8601_timestamp("timestamp"), '%Y') as "year",
+         date_format(from_iso8601_timestamp("timestamp"), '%m') as "month"
+      from bfd_cw_export.api_requests_no_partitions
+      --where year(from_iso8601_timestamp("timestamp")) = 2019
+      --where year(from_iso8601_timestamp("timestamp")) = 2020
+      --where year(from_iso8601_timestamp("timestamp")) = 2021
+      --where year(from_iso8601_timestamp("timestamp")) = 2022
+        --and month(from_iso8601_timestamp("timestamp")) <= 8
+      where year(from_iso8601_timestamp("timestamp")) = 2022
+         and month(from_iso8601_timestamp("timestamp")) = 8
+      ```
+   
+7. Verify the load.
+   1. Select a sample of the data and inspect the most important columns: timestamp, mdc_bene_id, mdc_http* to ensure
+      that the columns are populated sensibly. Note that many of the other columns are only sparsely populated.
```

Comment:
Not sure if its in the scope of this runbook, but what do we do if this got messed up and isnt populated sensibly?

Reply:
There are many reasons why that might happen but I think they all come down to -- the runbook was not followed correctly or the runbook is wrong or misleading in some way. I'm considering stuff like that to be out of scope here as it would be difficult to anticipate the ways that it could go wrong.